cpp/ql/lib/semmle/code/cpp/commons/Printf.qll

```diff
@@ -1046,39 +1060,64 @@ class FormatLiteral extends Literal {
         or
         this.getConversionChar(n).toLowerCase() = ["d", "i"] and
         // e.g. -2^31 = "-2147483648"
-        exists(int sizeBits |
-          sizeBits =
-            min(int bits |
-              bits = this.getIntegralDisplayType(n).getSize() * 8
-              or
-              exists(IntegralType t |
-                t = this.getUse().getConversionArgument(n).getType().getUnderlyingType()
-              |
-                t.isSigned() and bits = t.getSize() * 8
-              )
-            ) and
-          len = 1 + ((sizeBits - 1) / 10.0.log2()).ceil()
-          // this calculation is as %u (below) only we take out the sign bit (- 1) and allow a whole
-          // character for it to be expressed as '-'.
-        )
+        len =
+          min(float cand |
+            // The first case handles length sub-specifiers
+            // Subtract one in the exponent because one bit is for the sign.
+            // Add 1 to account for the possible sign in the output.
+            cand = 1 + lengthInBase10(2.pow(this.getIntegralDisplayType(n).getSize() * 8 - 1))
+            or
+            // The second case uses range analysis to deduce a length that's shorter than the length
+            // of the number -2^31.
+            exists(Expr arg, float lower |
+              arg = this.getUse().getConversionArgument(n) and
+              lower = lowerBound(arg.getFullyConverted())
+            |
+              cand =
+                max(int cand0 |
+                  (
+                    // Include the sign bit in the length of `lower` if it can be negative
+                    if lower < 0
+                    then cand0 = 1 + lengthInBase10(lower.abs())
+                    else cand0 = lengthInBase10(lower)
+                  )
+                  or
+                  // We don't care about the sign of `upper`: if `upper` is negative, then we know
+                  // `lower` is also (possibly more) negative, and thus its length will be greater.
```

Comment:
Do we always have a lower bound if we have an upper bound (and vice-versa).

Reply:
Good question! I _think_ so, but I'm not really sure. I'll investigate this more closely.

Reply:
I think this invariant holds in `SimpleRangeAnalysis`, but in the spirit of removing implicit assumptions, I've rewritten this piece of code in 3f0bfe1d752fe7c297f5fc7154d59cc888f716a4 to be slightly less smart.

Reply:
I prefer the new version. :+1: